File: Source/Fantasy.Jobs/Client/JobListener.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ServiceModel;
using Fantasy.Jobs.Management;
using System.Threading;
using Fantasy.ServiceModel;
using Fantasy.Tracking;

namespace Fantasy.Jobs.Client
{
    [CallbackBehavior(ConcurrencyMode = ConcurrencyMode.Reentrant)]
    public class JobListener : IJobEventHandler, IDisposable
    {

        public JobListener()
        {
            _refreshThread = ThreadFactory.CreateThread(this.Refresh);
            

            _listenedJobs = new Collection<Guid>();
            _listenedJobs.Inserted += new EventHandler<CollectionEventArgs<Guid>>(ListenedJobsInserted);
            _listenedJobs.Removed += new EventHandler<CollectionEventArgs<Guid>>(ListenedJobsRemoved);
            _listenedJobs.Cleared += new EventHandler(ListenedJobsCleared);
            
        }

        void ListenedJobsCleared(object sender, EventArgs e)
        {
            
        }

        void ListenedJobsRemoved(object sender, CollectionEventArgs<Guid> e)
        {
            
        }

        void ListenedJobsInserted(object sender, CollectionEventArgs<Guid> e)
        {
            
        }

        public void Start()
        {
            _refreshThread.Start();
        }


        private bool _connected;

        public bool Connected
        {
            get { return _connected; }
            set 
            {
                if (_connected != value)
                {
                    _connected = value;

                    if (_connected)
                    {
                        this.OnConnect(EventArgs.Empty);
                    }
                    else
                    {
                        this.OnDisconnect(EventArgs.Empty); 
                    }
                }
            }
        }

        public event EventHandler Connect;

        protected virtual void OnConnect(EventArgs e)
        {
            if (this.Connect != null)
            {
                this.Connect(this, e);
            }
        }


        public event EventHandler Disconnect;

        protected virtual void OnDisconnect(EventArgs e)
        {
            if (this.Disconnect != null)
            {
                this.Disconnect(this, e);
            }
        }


        public event EventHandler<JobListenerJobEventArgs> JobAdded;

        protected virtual void OnJobAdded(JobListenerJobEventArgs e)
        {
            if (this.JobAdded != null)
            {
                this.JobAdded(this, e);
            }
        }


        public event EventHandler<JobListenerJobEventArgs > JobStateChanged;

        protected virtual void OnJobStateChanged(JobListenerJobEventArgs  e)
        {
            if (this.JobStateChanged != null)
            {
                this.JobStateChanged(this, e);
            }
        }


        public event EventHandler<JobListenerTrackEventArgs> TrackCreated;

        protected virtual void OnTrackCreated(JobListenerTrackEventArgs e)
        {
            if (this.TrackCreated != null)
            {
                this.TrackCreated(this, e);
            }
        }

        public event EventHandler<JobListenerTrackEventArgs> TrackChanged;

        protected virtual void OnTrackChanged(JobListenerTrackEventArgs e)
        {
            if (this.TrackChanged != null)
            {
                this.TrackChanged(this, e);
            }
        }


        private Collection<Guid> _listenedJobs;

        public IList<Guid> ListenedJobs
        {
            get { return _listenedJobs; }
        }

       

        private object _syncRoot = new object();

        private Thread _refreshThread;
        private ClientRef<IJobEventService> _jobEvent;

        private TrackFactory _trackConnection;

        public long Version { get; private set; }

        private void NewVersion()
        {
            lock (this._syncRoot)
            {
                Version = DateTime.Now.Ticks;
            }
        }

        private List<JobMetaData> _jobs = new List<JobMetaData>();
        public JobMetaData[] JobMetaData
        {
            get
            {
                lock (this._syncRoot)
                {
                    return _jobs.ToArray();
                }
            }
        }

        private List<ITrackListener> _tracks = new List<ITrackListener>();
        public ITrackListener[] Tracks
        {
            get
            {
                lock (this._syncRoot)
                {
                    return _tracks.ToArray();
                }
            }
        }


        private void Refresh()
        {
            TimeSpan waitTime;
            _trackConnection = new TrackFactory();
            while (true)
            {

                try
                {
                    bool reload = false;

                    if (_jobEvent == null)
                    {
                        reload = true;
                        _jobEvent = ClientFactory.CreateDuplex<IJobEventService>(this);
                        _jobEvent.Client.Connect();
                    }
                    else
                    {
                        _jobEvent.Client.Echo();
                    }
                    if (reload)
                    {

                        using (ClientRef<IJobService> service = ClientFactory.Create<IJobService>())
                        {
                            foreach (Guid id in this._listenedJobs)
                            {

                                JobMetaData job = service.Client.FindJobById(id);
                                if (job != null)
                                {

                                    JobMetaData old = this._jobs.Find(j => j.Id == id);
                                    if (old == null)
                                    {
                                        this._jobs.Add(job);
                                        this.OnJobAdded(new JobListenerJobEventArgs(job));
                                    }
                                    else if (old.State != job.State)
                                    {
                                        this.OnJobStateChanged(new JobListenerJobEventArgs(job));
                                    }

                                    CreateOrDisposeTrack(job);
                                }
                            }
                        }

                        lock (_syncRoot)
                        {
                            this.NewVersion();
                        }
                    }
                    waitTime = new TimeSpan(0, 2, 0);
                    this.Connected = true;
                }
                catch (ThreadAbortException)
                {
                    waitTime = TimeSpan.MinValue; 
                    if (_jobEvent != null && _jobEvent.State == CommunicationState.Opened)
                    {
                        _jobEvent.Client.Disconnect();
                    }
                }
                catch
                {
                    _jobEvent = null;
                    lock (_syncRoot)
                    {
                        this.NewVersion();
                    }
                    waitTime = new TimeSpan(0, 0, 10);
                    this.Connected = false;
                }
                Thread.Sleep(waitTime);
            }
        }

        private void CreateOrDisposeTrack(JobMetaData job)
        {
            Guid id = job.Id;
            ITrackListener oldTrack = this._tracks.Find(t => t.Id == id);
            if (job.State == JobState.Running && oldTrack == null)
            {
                ITrackListener track = this._trackConnection.CreateTrackListener(id);
                track.Changed += new EventHandler<TrackChangedEventArgs>(TrackListenerChanged);
                this._tracks.Add(track);
                this.OnTrackCreated(new JobListenerTrackEventArgs(track, null));
            }
            else if (job.State != JobState.Running && oldTrack != null)
            {
                this._tracks.Remove(oldTrack);
                oldTrack.Changed -= new EventHandler<TrackChangedEventArgs>(TrackListenerChanged);
                oldTrack.Dispose();
            }
        }

        void TrackListenerChanged(object sender, TrackChangedEventArgs e)
        {
            this.OnTrackChanged(new JobListenerTrackEventArgs((ITrackListener)sender, e)); 
        }

      
        #region IJobEventHandler Members

        void IJobEventHandler.Added(JobMetaData job)
        {
            lock (_syncRoot)
            {
                if (this._listenedJobs.IndexOf(job.Id) >= 0 && this._jobs.Find(j => j.Id == job.Id) == null)
                {
                    this._jobs.Add(job);
                    this.OnJobAdded(new JobListenerJobEventArgs(job));
                }
            }
        }

        void IJobEventHandler.Changed(JobMetaData job)
        {
            lock (_syncRoot)
            {
                if (this._listenedJobs.IndexOf(job.Id) >= 0)
                {
                    int idx = this._jobs.FindIndex(j => j.Id == job.Id);
                    if (idx < 0)
                    {
                        this._jobs.Add(job);
                        this.OnJobAdded(new JobListenerJobEventArgs(job));
                    }
                    else
                    {
                        this._jobs[idx] = job;
                        this.OnJobStateChanged(new JobListenerJobEventArgs(job)); 
                    }

                    CreateOrDisposeTrack(job);
                }
                this.NewVersion();
            }
        }

        void IJobEventHandler.Echo()
        {

        }

        #endregion

        #region IDisposable Members

        public void Dispose()
        {
            this._refreshThread.Abort();
        }

        #endregion
    }


    public class JobListenerJobEventArgs : EventArgs
    {
    
        internal JobListenerJobEventArgs(JobMetaData job)
        {
            this.Job = job;
        }

        public JobMetaData Job { get; private set; }
    }

    public class JobListenerTrackEventArgs : EventArgs
    {

        internal JobListenerTrackEventArgs(ITrackListener track, TrackChangedEventArgs e)
        {
            this.Track = track;
            if (e != null)
            {
                Name = e.Name;
                OldValue = e.OldValue;
                NewValue = e.NewValue;
            }
        }

        public ITrackListener Track { get; private set; }

        public Guid Id
        {
            get
            {
                return Track.Id;
            }
        }

        public string Name { get; private set; }
        public object OldValue { get; private set; }
        public object NewValue { get; private set; }
    }
}

File: branches/Stheno/AutoConverted/Fantasy.Jobs/Client/JobListener.java
package Fantasy.Jobs.Client;

import Fantasy.Jobs.Management.*;
import Fantasy.ServiceModel.*;
import Fantasy.Tracking.*;

//C# TO JAVA CONVERTER TODO TASK: Java annotations will not correspond to .NET attributes:
//[CallbackBehavior(ConcurrencyMode = ConcurrencyMode.Reentrant)]
public class JobListener implements IJobEventHandler, IDisposable
{

	public JobListener()
	{
		_refreshThread = ThreadFactory.CreateThread(this.Refresh);


		_listenedJobs = new Collection<Guid>();
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
		_listenedJobs.Inserted += new EventHandler<CollectionEventArgs<Guid>>(ListenedJobsInserted);
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
		_listenedJobs.Removed += new EventHandler<CollectionEventArgs<Guid>>(ListenedJobsRemoved);
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
		_listenedJobs.Cleared += new EventHandler(ListenedJobsCleared);

	}

	private void ListenedJobsCleared(Object sender, EventArgs e)
	{

	}

	private void ListenedJobsRemoved(Object sender, CollectionEventArgs<Guid> e)
	{

	}

	private void ListenedJobsInserted(Object sender, CollectionEventArgs<Guid> e)
	{

	}

	public final void Start()
	{
		_refreshThread.start();
	}


	private boolean _connected;

	public final boolean getConnected()
	{
		return _connected;
	}
	public final void setConnected(boolean value)
	{
		if (_connected != value)
		{
			_connected = value;

			if (_connected)
			{
				this.OnConnect(EventArgs.Empty);
			}
			else
			{
				this.OnDisconnect(EventArgs.Empty);
			}
		}
	}

//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler Connect;

	protected void OnConnect(EventArgs e)
	{
		if (this.Connect != null)
		{
			this.Connect(this, e);
		}
	}


//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler Disconnect;

	protected void OnDisconnect(EventArgs e)
	{
		if (this.Disconnect != null)
		{
			this.Disconnect(this, e);
		}
	}


//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler<JobListenerJobEventArgs> JobAdded;

	protected void OnJobAdded(JobListenerJobEventArgs e)
	{
		if (this.JobAdded != null)
		{
			this.JobAdded(this, e);
		}
	}


//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler<JobListenerJobEventArgs > JobStateChanged;

	protected void OnJobStateChanged(JobListenerJobEventArgs e)
	{
		if (this.JobStateChanged != null)
		{
			this.JobStateChanged(this, e);
		}
	}


//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler<JobListenerTrackEventArgs> TrackCreated;

	protected void OnTrackCreated(JobListenerTrackEventArgs e)
	{
		if (this.TrackCreated != null)
		{
			this.TrackCreated(this, e);
		}
	}

//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler<JobListenerTrackEventArgs> TrackChanged;

	protected void OnTrackChanged(JobListenerTrackEventArgs e)
	{
		if (this.TrackChanged != null)
		{
			this.TrackChanged(this, e);
		}
	}


	private Collection<Guid> _listenedJobs;

	public final java.util.List<Guid> getListenedJobs()
	{
		return _listenedJobs;
	}



	private Object _syncRoot = new Object();

	private Thread _refreshThread;
	private ClientRef<IJobEventService> _jobEvent;

	private TrackFactory _trackConnection;

	private long privateVersion;
	public final long getVersion()
	{
		return privateVersion;
	}
	private void setVersion(long value)
	{
		privateVersion = value;
	}

	private void NewVersion()
	{
		synchronized (this._syncRoot)
		{
			setVersion(new java.util.Date().Ticks);
		}
	}

	private java.util.ArrayList<JobMetaData> _jobs = new java.util.ArrayList<JobMetaData>();
	public final JobMetaData[] getJobMetaData()
	{
		synchronized (this._syncRoot)
		{
			return _jobs.toArray(new JobMetaData[]{});
		}
	}

	private java.util.ArrayList<ITrackListener> _tracks = new java.util.ArrayList<ITrackListener>();
	public final ITrackListener[] getTracks()
	{
		synchronized (this._syncRoot)
		{
			return _tracks.toArray(new ITrackListener[]{});
		}
	}


	private void Refresh()
	{
		TimeSpan waitTime = new TimeSpan();
		_trackConnection = new TrackFactory();
		while (true)
		{

			try
			{
				boolean reload = false;

				if (_jobEvent == null)
				{
					reload = true;
					_jobEvent = ClientFactory.<IJobEventService>CreateDuplex(this);
					_jobEvent.Client.Connect();
				}
				else
				{
					_jobEvent.Client.Echo();
				}
				if (reload)
				{

//C# TO JAVA CONVERTER NOTE: The following 'using' block is replaced by its Java equivalent:
//					using (ClientRef<IJobService> service = ClientFactory.Create<IJobService>())
					ClientRef<IJobService> service = ClientFactory.<IJobService>Create();
					try
					{
						for (Guid id : this._listenedJobs)
						{

							JobMetaData job = service.Client.FindJobById(id);
							if (job != null)
							{

//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
								JobMetaData old = this._jobs.Find(j => id.equals(j.Id));
								if (old == null)
								{
									this._jobs.add(job);
									this.OnJobAdded(new JobListenerJobEventArgs(job));
								}
								else if (old.getState() != job.getState())
								{
									this.OnJobStateChanged(new JobListenerJobEventArgs(job));
								}

								CreateOrDisposeTrack(job);
							}
						}
					}
					finally
					{
					}

					synchronized (_syncRoot)
					{
						this.NewVersion();
					}
				}
				waitTime = new TimeSpan(0, 2, 0);
				this.setConnected(true);
			}
			catch (ThreadAbortException e)
			{
				waitTime = TimeSpan.getMinValue();
				if (_jobEvent != null && _jobEvent.State == CommunicationState.Opened)
				{
					_jobEvent.Client.Disconnect();
				}
			}
			catch (java.lang.Exception e2)
			{
				_jobEvent = null;
				synchronized (_syncRoot)
				{
					this.NewVersion();
				}
				waitTime = new TimeSpan(0, 0, 10);
				this.setConnected(false);
			}
			Thread.sleep(waitTime);
		}
	}

	private void CreateOrDisposeTrack(JobMetaData job)
	{
		Guid id = job.getId();
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
		ITrackListener oldTrack = this._tracks.Find(t => id.equals(t.Id));
		if (job.getState() == JobState.Running && oldTrack == null)
		{
			ITrackListener track = this._trackConnection.CreateTrackListener(id);
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
			track.Changed += new EventHandler<TrackChangedEventArgs>(TrackListenerChanged);
			this._tracks.add(track);
			this.OnTrackCreated(new JobListenerTrackEventArgs(track, null));
		}
		else if (job.getState() != JobState.Running && oldTrack != null)
		{
			this._tracks.remove(oldTrack);
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
			oldTrack.Changed -= new EventHandler<TrackChangedEventArgs>(TrackListenerChanged);
			oldTrack.dispose();
		}
	}

	private void TrackListenerChanged(Object sender, TrackChangedEventArgs e)
	{
		this.OnTrackChanged(new JobListenerTrackEventArgs((ITrackListener)sender, e));
	}


//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#region IJobEventHandler Members

	private void Added(JobMetaData job)
	{
		synchronized (_syncRoot)
		{
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			if (this._listenedJobs.indexOf(job.getId()) >= 0 && this._jobs.Find(j => job.getId().equals(j.Id)) == null)
			{
				this._jobs.add(job);
				this.OnJobAdded(new JobListenerJobEventArgs(job));
			}
		}
	}

	private void Changed(JobMetaData job)
	{
		synchronized (_syncRoot)
		{
			if (this._listenedJobs.indexOf(job.getId()) >= 0)
			{
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
				int idx = this._jobs.FindIndex(j => job.getId().equals(j.Id));
				if (idx < 0)
				{
					this._jobs.add(job);
					this.OnJobAdded(new JobListenerJobEventArgs(job));
				}
				else
				{
					this._jobs.set(idx, job);
					this.OnJobStateChanged(new JobListenerJobEventArgs(job));
				}

				CreateOrDisposeTrack(job);
			}
			this.NewVersion();
		}
	}

	private void Echo()
	{

	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#endregion

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#region IDisposable Members

	public final void dispose()
	{
		this._refreshThread.stop();
	}

//C# TO JAVA CONVERTER TODO TASK: There is no preprocessor in Java:
		///#endregion
}